internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,200 @@
+// It supports optional caching, cache removal, synchronization, and context cancellation.
+// Given job options, input items, an optional cache lookup/removal function, and a work function,
+// RunJob() processes each item in a parallel pool.
+// Returns the result list, number of remaining items, total run time, and an error.
+// If the context is canceled, returns nil, 0, 0, nil.
+func RunJob[InputType any, OutputType any](options *JobOptions, workItems []InputType, cacheFunc func([]InputType) ([]OutputType, []InputType, error), removeCacheFunc func([]InputType) error, workFunc func(InputType) (OutputType, int64, error)) ([]OutputType, int, int64, error) {
+	if workFunc == nil {
+		return nil, len(workItems), 0, fmt.Errorf("Cannot run job with a nil work function.")
+	}
+
+	err := options.Validate()
+	if err != nil {
+		return nil, len(workItems), 0, fmt.Errorf("Failed to validate job options: '%v'.", err)
+	}
+
+	completeItems := make([]OutputType, 0)
+	remainingItems := workItems
+
+	// If we are not overwriting the cache, query for cached results.
+	if !options.OverwriteCache && cacheFunc != nil {
+		completeItems, remainingItems, err = cacheFunc(workItems)
+		if err != nil {
+			return nil, len(workItems), 0, err
+		}
+	}
+
+	runTime := int64(0)
+	results := make([]OutputType, 0)
+
+	if options.WaitForCompletion {
+		results, runTime, err = runJob(options, remainingItems, cacheFunc, removeCacheFunc, workFunc)
+		if err != nil {
+			return nil, len(workItems), 0, err
+		}
+
+		// If the context was canceled during execution, return immediately.
+		if options.Context.Err() != nil {
+			return nil, 0, 0, nil
+		}
+
+		completeItems = append(completeItems, results...)
+		remainingItems = nil
+	} else {
+		go func() {
+			_, _, err := runJob(options, remainingItems, cacheFunc, removeCacheFunc, workFunc)
+			if err != nil {
+				log.Error("Failure while running asynchronous job.")
+			}
+		}()
+	}
+
+	return completeItems, len(remainingItems), runTime, nil
+}
+
+func runJob[InputType any, OutputType any](options *JobOptions, workItems []InputType, cacheFunc func([]InputType) ([]OutputType, []InputType, error), removeCacheFunc func([]InputType) error, workFunc func(InputType) (OutputType, int64, error)) ([]OutputType, int64, error) {
+	if len(workItems) == 0 {
+		return nil, 0, nil
+	}
+
+	err := options.Validate()
+	if err != nil {
+		return []OutputType{}, 0, fmt.Errorf("Failed to validate job options: '%v'.", err)
+	}
+
+	noLockWait := lockmanager.Lock(options.LockKey)
+	defer lockmanager.Unlock(options.LockKey)
+
+	// The context has been canceled while waiting for a lock, abandon this job.
+	if options.Context.Err() != nil {
+		return nil, 0, nil
+	}
+
+	results := make([]OutputType, 0, len(workItems))
+
+	// If we had to wait for the lock, then check again for cached results.
+	// If there are multiple requests queued up,
+	// it will be faster to do a bulk check for cached results instead of checking each one individually.
```

Comment:
This sentence no longer makes sense in this context.
For analysis, we would check caches in bulk and on an individual level,
but here we only check in bulk.